pycozmo/window.py

```diff
@@ -98,44 +99,67 @@ def __init__(self, seq_bits: int, size: Optional[int] = None) -> None:
         self.next_seq = 1
         self.window = [SendWindowSlot() for _ in range(self.size)]
+        self.lock = Lock()
 
     def is_out_of_order(self, seq: int) -> bool:
-        if self.is_empty():
-            res = True
-        elif self.expected_seq > self.next_seq:
-            res = self.expected_seq > seq >= self.next_seq
-        else:
-            res = seq < self.expected_seq or seq >= self.next_seq
-        return res
+        with self.lock:
+            if self.is_empty():
+                res = True
+            elif self.expected_seq > self.next_seq:
+                res = self.expected_seq > seq >= self.next_seq
+            else:
+                res = seq < self.expected_seq or seq >= self.next_seq
+            return res
 
     def is_empty(self) -> bool:
-        res = self.expected_seq == self.next_seq
-        return res
+        with self.lock:
+            res = self.expected_seq == self.next_seq
+            return res
 
     def is_full(self) -> bool:
-        if self.expected_seq > self.next_seq:
-            res = (self.next_seq + self.max_seq - self.expected_seq) >= self.size
-        else:
-            res = (self.next_seq - self.expected_seq) >= self.size
-        return res
-
-    def put(self, data: Any) -> int:
-        seq = self.next_seq
-        self.next_seq = (self.next_seq + 1) % self.max_seq
-        self.window[seq % self.size].set(seq, data)
-        return seq
+        with self.lock:
+            if self.expected_seq > self.next_seq:
+                res = (self.next_seq + self.max_seq - self.expected_seq) >= self.size
+            else:
+                res = (self.next_seq - self.expected_seq) >= self.size
+            return res
+
+    def put(self, data: Any) -> tuple(int,list):
+        with self.lock:
+            seq = self.next_seq
+            self.next_seq = (self.next_seq + 1) % self.max_seq
+            self.window[seq % self.size].set(seq, data)
+            return self.expected_seq, self.window[seq:self.expected_seq]
 
     def pop(self) -> None:
-        self.window[self.expected_seq % self.size].reset()
-        self.expected_seq = (self.expected_seq + 1) % self.max_seq
-        self.last_seq = (self.last_seq + 1) % self.max_seq
+        with self.lock:
+            self.window[self.expected_seq % self.size].reset()
+            self.expected_seq = (self.expected_seq + 1) % self.max_seq
+            self.last_seq = (self.last_seq + 1) % self.max_seq
+
+    def acknowledge(self, seq: int):
+        with self.lock:
+            if self.is_out_of_order(seq):
+                pass
+            else:
+                if seq > self.expected_seq:
+                    for frame in self.window[self.expected_seq:seq]:
+                        frame.reset()
+                else:
+                    for frame in self.window[self.expected_seq:]:
+                        frame.reset()
+                    for frame in self.window[:seq]:
+                        frame.reset()
```

Comment:
You are right, something like this?
```
for i in range(seq + self.expected_seq - self.size):
    self.window[(i+self.expected_seq)%self.size].reset()
```
This could substitute lines 145 to 152 actually, much better, thanks!